File: docs/palindromes-in-peg/scaffold_window_stream.py
"""Bounded head motion over blocks of the original, unexpanded input.

During one input round a head is an old head origin plus a finite offset.
Blocks have length B, greater than the round's movement bound. Only the old
block and its two neighbors can be reached. Their zipper states are prepared
once, and only the selected final state is saved. Intermediate movements and
copies allocate no cells. Input blocks contain actual input-node addresses.

All input characters must belong to this stream, in consecutive transitions.
Frozen views, result stacks, and the complete PAL compiler are separate work.
"""
from copy import copy

from scaffold_circuit import (Circuit, Value, Ref, PREVIOUS, NEW, EMPTY, TRUE, FALSE,
                              choose, choose_pointer, conjunction as AND,
                              disjunction as OR, neg)
from scaffold_circuit_structs import Stack, StackPool, Queue
from scaffold_window_counter import Bits
from symbolic_sca2peg import pointer, edge


def clone_counter(source):
  result = copy(source)
  result.pos, result.neg = copy(source.pos), copy(source.neg)
  return result


def clone_queue(source):
  result = copy(source)
  result.s = {name: copy(stack) for name, stack in source.s.items()}
  result.m, result.c = clone_counter(source.m), clone_counter(source.c)
  return result


class BlockState:
  def __init__(self, bank, name):
    self.bank, self.name, self.circuit = bank, name, bank.circuit
    self.focus_key, self.live_key = name + ".block_end", name + ".live"
    self.focus = self.circuit.get_ref(PREVIOUS, self.focus_key)
    self.live = self.circuit.get(PREVIOUS, self.live_key, (False, True), True).eq(True)
    self.left = Stack(bank.cells, name + ".left")
    self.right = Stack(bank.cells, name + ".right")
    self.queue = Queue(bank.queue_cells, bank.counter_cells, name + ".incoming")

  def clone(self):
    result = copy(self)
    result.left, result.right = copy(self.left), copy(self.right)
    result.queue = clone_queue(self.queue)
    return result

  def complete_block(self, completed):
    self.queue.push(NEW, AND(completed, neg(self.live)))
    self.queue.work(completed)
    self.focus = Ref.select(AND(completed, self.live), NEW, self.focus)
    self.live = AND(self.live, neg(completed))

  def move_left(self, enabled):
    # A live block is not pushed. If it later completes while this head is
    # elsewhere, completion enqueues its actual endpoint. Before completion,
    # an empty right side already denotes that same live block.
    self.right.push(self.focus, enabled=AND(enabled, neg(self.live)))
    focus, _ = self.left.pop(enabled)
    self.focus = Ref.select(enabled, focus, self.focus)
    self.live = AND(self.live, neg(enabled))

  def move_right(self, enabled):
    self.left.push(self.focus, enabled=enabled)
    stacked, queued = neg(self.right.empty()), neg(self.queue.empty())
    from_queue = AND(enabled, neg(stacked), queued)
    a, _ = self.right.pop(AND(enabled, stacked))
    b = self.queue.pop(from_queue)
    self.queue.work(from_queue)
    self.focus = Ref.select(enabled, Ref.select(stacked, a, b), self.focus)
    self.live = choose(enabled, AND(neg(stacked), neg(queued)), self.live)

  def copy_from(self, other, enabled):
    self.focus = Ref.select(enabled, other.focus, self.focus)
    self.live = choose(enabled, other.live, self.live)
    self.left.copy_from(other.left, enabled)
    self.right.copy_from(other.right, enabled)
    self.queue.copy_from(other.queue, enabled)

  def finalize(self):
    self.circuit.put_ref(self.focus_key, self.focus)
    self.circuit.put(self.live_key,
      Value.select(self.live, Value.constant(True), Value.constant(False)))
    self.left.finalize()
    self.right.finalize()
    self.queue.finalize()


class WindowHead:
  def __init__(self, bank, name):
    self.bank, self.name = bank, name
    self.origin = Value.constant(name).recode(bank.names)
    low = Bits.load(bank.circuit, name + ".offset", bank.width)
    self.offset = Bits((*low.bits, FALSE, FALSE))
    self.radius = 0

  def variants(self, offset=None):
    offset = self.offset if offset is None else offset
    high, sign = offset.bits[-2:]
    directions = {-1: AND(high, sign), 0: AND(neg(high), neg(sign)),
                   1: AND(high, neg(sign))}
    for name in self.bank.names:
      origin = self.origin.eq(name)
      if origin == FALSE:
        continue
      for direction, guard in directions.items():
        state, valid = self.bank.neighbors[name][direction]
        yield state, AND(origin, guard), valid

  def describe(self, offset=None):
    endpoint, live, valid = EMPTY, FALSE, FALSE
    for state, guard, permitted in self.variants(offset):
      endpoint = Ref.select(guard, state.focus, endpoint)
      live = OR(live, AND(guard, state.live))
      valid = OR(valid, AND(guard, permitted))
    return endpoint, live, valid

  def location_flags(self, offset=None):
    # Availability does not need an endpoint. Avoid constructing discarded
    # pointer-selection DAGs for every prospective fixed movement.
    live, valid = FALSE, FALSE
    for state, guard, permitted in self.variants(offset):
      live = OR(live, AND(guard, state.live))
      valid = OR(valid, AND(guard, permitted))
    return live, valid

  def available(self):
    live, valid = self.location_flags()
    low = Bits(self.offset.bits[:self.bank.width])
    return AND(valid, OR(neg(live), low.unsigned_less(self.bank.phase)))

  def can_move(self, amount):
    candidate = self.offset.add(amount)
    live, valid = self.location_flags(candidate)
    low = Bits(candidate.bits[:self.bank.width])
    return AND(valid, OR(neg(live), low.unsigned_less(self.bank.phase), low.equal(self.bank.phase)))

  def move(self, amount, enabled=TRUE):
    if type(amount) is not int:
      raise ValueError("head motion must have a fixed finite distance")
    if enabled == FALSE or amount == 0:
      return
    if self.radius + abs(amount) >= self.bank.base:
      raise ValueError("head lineage exceeds the declared finite window")
    self.bank.circuit.require(self.can_move(amount), enabled)
    self.offset = Bits.select(enabled, self.offset.add(amount), self.offset)
    self.radius += abs(amount)

  def copy_from(self, other, enabled=TRUE):
    if other.bank is not self.bank:
      raise ValueError("head copies require a common stream")
    self.origin = Value.select(enabled, other.origin, self.origin)
    self.offset = Bits.select(enabled, other.offset, self.offset)
    self.radius = other.radius if enabled == TRUE else max(self.radius, other.radius)

  def move_selected(self, amount, enabled=TRUE):
    bound = max(map(abs, amount.domain))
    if self.radius + bound >= self.bank.base:
      raise ValueError("selected head motion exceeds the prepared finite window")
    before, after = self.offset, self.offset
    for delta in amount.domain:
      if not delta: continue
      guard = AND(enabled, amount.eq(delta))
      self.bank.circuit.require(self.can_move(delta), guard)
      after = Bits.select(guard, before.add(delta), after)
    self.offset, self.radius = after, self.radius + bound

  def read(self, enabled=TRUE):
    c, bank = self.bank.circuit, self.bank
    available = self.available()
    c.require(available, enabled)
    endpoint, live, _ = self.describe()
    low = Bits(self.offset.bits[:bank.width])
    full_distance = Bits(tuple(neg(bit) for bit in low.bits))
    live_distance = bank.phase.plus(low.negated()).add(-1)
    distance = Bits.select(live, live_distance, full_distance)
    endpoint = Ref.select(live, NEW, endpoint)
    for bit, enabled_bit in enumerate(distance.bits):
      endpoint = Ref.select(enabled_bit, bank.hop(endpoint, 1 << bit), endpoint)
    value = c.get(endpoint, bank.input_key, c.alphabet, c.alphabet[0])
    return Value.select(available, value, Value.constant(None))

  def read_relative(self, amount, enabled=TRUE):
    if self.radius + abs(amount) >= self.bank.base:
      raise ValueError("relative read exceeds the prepared head window")
    temporary = copy(self)
    temporary.offset = self.offset.add(amount)
    return temporary.read(enabled)

  def finalize(self):
    target = self.bank.states[self.name]
    for source, guard, valid in self.variants():
      self.bank.circuit.require(valid, guard)
      target.copy_from(source, guard)
    Bits(self.offset.bits[:self.bank.width]).store(self.bank.circuit, self.name + ".offset")
    target.finalize()


class WindowStream:
  def __init__(self, circuit, names, radius, prefix="stream"):
    if type(radius) is not int or radius < 0:
      raise ValueError("a nonnegative finite movement radius is required")
    self.circuit, self.names, self.prefix = circuit, tuple(names), prefix
    if not self.names or len(set(self.names)) != len(self.names):
      raise ValueError("distinct finite head names required")
    self.width = max(1, radius.bit_length())
    self.base = 1 << self.width
    self.phase_key, self.input_key = prefix + ".phase", prefix + ".input"
    old_phase = Bits.load(circuit, self.phase_key, self.width)
    completed = AND(*old_phase.bits)
    self.phase = old_phase.add(1)
    circuit.put(self.input_key, circuit.input(), circuit.alphabet, circuit.alphabet[0])
    self.cells = StackPool(circuit, {name + side: 1 for name in self.names for side in (".left", ".right")})
    def pool(layout):
      return StackPool(circuit, {f"{name}.incoming.{role}": count for name in self.names
                                for role, count in layout})
    front, rear, reverse = pool((("Br", 12),)), pool((("B", 1), ("B2", 1))), pool((("Fr", 6),))
    self.queue_cells = {"F": front, "WF": front, "Br": front,
                        "B": rear, "WB": rear, "B2": rear, "Fr": reverse}
    self.counter_cells = {"m": {"pos": pool((("m.pos", 6),)), "neg": pool((("m.neg", 8),))},
                          "c": {"pos": pool((("c.pos", 12),)), "neg": pool((("c.neg", 2),))}}
    self.states = {name: BlockState(self, name) for name in self.names}
    self.neighbors, self.jumps = {}, {}
    for name, state in self.states.items():
      state.complete_block(completed)
      center = state.clone()
      before, after = center.clone(), center.clone()
      can_left, can_right = neg(center.left.empty()), neg(center.live)
      before.move_left(can_left)
      after.move_right(can_right)
      self.neighbors[name] = {-1: (before, can_left), 0: (center, TRUE), 1: (after, can_right)}
    self.heads = {name: WindowHead(self, name) for name in self.names}

  def jump(self, distance):
    """An original-input predecessor by a fixed distance, in O(log B) fields."""
    if distance < 1:
      raise ValueError("positive original-input hop required")
    if distance not in self.jumps:
      name = f"{self.prefix}.back{distance}"
      self.jumps[distance] = name
      if distance == 1:
        path = ()
      elif distance % 2:
        path = (self.jump(distance // 2),) * 2
      else:
        half = distance // 2
        path = (self.jump(half),) + ((self.jump(half - 1),) if half > 1 else ())
      self.circuit.pointers[name] = pointer(path)
    return self.jumps[distance]

  def hop(self, target, distance):
    from_new = pointer(()) if distance == 1 else pointer((self.jump(distance - 1),))
    return Ref(FALSE, choose_pointer(target.new, from_new, edge(target.prior, self.jump(distance))))

  def select_head(self, which):
    cases = [(which.eq(name), head) for name, head in self.heads.items() if which.eq(name) != FALSE]
    result = object.__new__(WindowHead)
    result.bank, result.name = self, None
    result.origin = Value.encoded(self.names,
      (OR(*(AND(guard, head.origin.bits[bit]) for guard, head in cases))
       for bit in range((len(self.names) - 1).bit_length())),
      OR(*(guard for guard, _ in cases)))
    result.offset = Bits(tuple(OR(*(AND(guard, head.offset.bits[bit]) for guard, head in cases))
                              for bit in range(self.width + 2)))
    result.radius = max((head.radius for _, head in cases), default=0)
    return result

  def finalize(self):
    for head in self.heads.values():
      head.finalize()
    self.phase.store(self.circuit, self.phase_key)


def fixture(quantum=3):
  circuit = Circuit("abcdefgh")
  stream = WindowStream(circuit, ("x", "y"), 2 * quantum + 2)
  x, y = (stream.heads[name] for name in stream.names)
  char = circuit.input()
  for _ in range(quantum):
    x.move(1, AND(char.eq("a"), x.can_move(1)))
    x.move(-1, AND(char.eq("b"), x.can_move(-1)))
  y.copy_from(x, char.eq("c"))
  x.copy_from(y, char.eq("d"))
  y.move(1, AND(char.eq("e"), y.can_move(1)))
  y.move(-1, AND(char.eq("f"), y.can_move(-1)))
  observations = {}
  for head in (x, y):
    value = head.read(head.available())
    key = head.name + ".read"
    circuit.put(key, value, (*circuit.alphabet, None), None)
    observations[head.name] = key
  answer = x.read(x.available()).eq("a")
  stream.finalize()
  return circuit, stream, observations, circuit.machine(answer)

File: scala/pal/src/main/scala/pal/ScaffoldWindowStream.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable
import Expr.{TRUE, FALSE, pointer, edge}
import Ref.{PREVIOUS, NEW, EMPTY}
import ScaffoldCircuit.{conjunction as AND, disjunction as OR, choose, choosePointer, neg}

object ScaffoldWindowStream {
  private[pal] def cloneStack(source: Stack): Stack = {
    val result = new Stack(source.pool, source.name, Some(source.allocationName))
    result.top = source.top
    result.tag = source.tag
    result
  }
  def cloneCounter(source: Counter): Counter = {
    val result = new Counter(Map("pos" -> source.pos.pool, "neg" -> source.negativeStack.pool), source.name, Some(source.pos.allocationName))
    result.positiveStack = cloneStack(source.pos)
    result.negativeStack = cloneStack(source.negativeStack)
    result
  }
  def cloneQueue(source: Queue): Queue = {
    val pools = source.stacks.view.mapValues(_.pool).toMap
    val counterPools = Map("m" -> Map("pos" -> source.m.pos.pool, "neg" -> source.m.negativeStack.pool),
      "c" -> Map("pos" -> source.c.pos.pool, "neg" -> source.c.negativeStack.pool))
    val result = new Queue(pools, counterPools, source.name, source.sharedSlots)
    // Each new queue owns fresh stack objects, but shares physical pools.
    for ((name, stack) <- source.stacks) {
      result.stacks(name).top = stack.top
      result.stacks(name).tag = stack.tag
    }
    for ((target, prior) <- Vector(result.m -> source.m, result.c -> source.c)) {
      target.positiveStack = cloneStack(prior.pos)
      target.negativeStack = cloneStack(prior.negativeStack)
    }
    result.phase = source.phase
    result
  }
  def fixture(quantum: Int = 3): (Circuit, WindowStream, VectorMap[String, String], Scaffold) = {
    val circuit = new Circuit("abcdefgh")
    val stream = new WindowStream(circuit, Vector("x", "y"), 2 * quantum + 2)
    val x = stream.heads("x")
    val y = stream.heads("y")
    val char = circuit.input()
    for (_ <- 0 until quantum) {
      x.move(1, AND(char.eqTo('a'), x.canMove(1)))
      x.move(-1, AND(char.eqTo('b'), x.canMove(-1)))
    }
    y.copyFrom(x, char.eqTo('c'))
    x.copyFrom(y, char.eqTo('d'))
    y.move(1, AND(char.eqTo('e'), y.canMove(1)))
    y.move(-1, AND(char.eqTo('f'), y.canMove(-1)))
    val observations = VectorMap.from(Vector(x, y).map { head =>
      val value = head.read(head.available())
      val key = head.name + ".read"
      circuit.put(key, value, circuit.alphabet :+ None, None)
      head.name -> key
    })
    val answer = x.read(x.available()).eqTo('a')
    stream.commit()
    (circuit, stream, observations, circuit.machine(answer))
  }
}

/** A block zipper snapshot; all clones share storage but own evolving roots. */
final class BlockState(val bank: WindowStream, val name: String) {
  val circuit: Circuit = bank.circuit
  val focusKey: String = name + ".block_end"
  val liveKey: String = name + ".live"
  var focus: Ref = circuit.getRef(PREVIOUS, focusKey)
  var live: Expr = circuit.get(PREVIOUS, liveKey, Vector(false, true), true).eqTo(true)
  var left: Stack = new Stack(bank.cells, name + ".left")
  var right: Stack = new Stack(bank.cells, name + ".right")
  var queue: Queue = new Queue(bank.queueCells, bank.counterCells, name + ".incoming")
  override def clone(): BlockState = {
    val result = new BlockState(bank, name)
    result.focus = focus
    result.live = live
    result.left = ScaffoldWindowStream.cloneStack(left)
    result.right = ScaffoldWindowStream.cloneStack(right)
    result.queue = ScaffoldWindowStream.cloneQueue(queue)
    result
  }
  def completeBlock(completed: Expr): Unit = {
    queue.push(NEW, AND(completed, neg(live)))
    queue.work(completed)
    focus = Ref.select(AND(completed, live), NEW, focus)
    live = AND(live, neg(completed))
  }
  def moveLeft(enabled: Expr): Unit = {
    // A live block is not pushed: later completion queues its actual endpoint.
    right.push(focus, enabled = AND(enabled, neg(live)))
    val (prior, _) = left.pop(enabled)
    focus = Ref.select(enabled, prior, focus)
    live = AND(live, neg(enabled))
  }
  def moveRight(enabled: Expr): Unit = {
    left.push(focus, enabled = enabled)
    val stacked = neg(right.empty())
    val queued = neg(queue.empty())
    val fromQueue = AND(enabled, neg(stacked), queued)
    val (a, _) = right.pop(AND(enabled, stacked))
    val b = queue.pop(fromQueue)
    queue.work(fromQueue)
    focus = Ref.select(enabled, Ref.select(stacked, a, b), focus)
    live = choose(enabled, AND(neg(stacked), neg(queued)), live)
  }
  def copyFrom(other: BlockState, enabled: Expr): Unit = {
    focus = Ref.select(enabled, other.focus, focus)
    live = choose(enabled, other.live, live)
    left.copyFrom(other.left, enabled)
    right.copyFrom(other.right, enabled)
    queue.copyFrom(other.queue, enabled)
  }
  def commit(): Unit = {
    circuit.putRef(focusKey, focus)
    circuit.put(liveKey, Value.select(live, Value.constant(true), Value.constant(false)))
    left.commit()
    right.commit()
    queue.commit()
  }
}

final class WindowHead private[pal] (val bank: WindowStream, val name: String,
                                    var origin: Value[String], var offset: Bits, var radius: BigInt) {
  def this(bank: WindowStream, name: String) = {
    this(bank, name, Value.constant(name).recode(bank.names),
      Bits(Bits.load(bank.circuit, name + ".offset", bank.width).bits ++ Vector(FALSE, FALSE)), BigInt(0))
  }
  def variants(): Vector[(BlockState, Expr, Expr)] = variants(offset)
  def variants(offset: Bits): Vector[(BlockState, Expr, Expr)] = {
    val high = offset.bits(offset.bits.size - 2)
    val sign = offset.bits.last
    val directions = Vector(-1 -> AND(high, sign), 0 -> AND(neg(high), neg(sign)), 1 -> AND(high, neg(sign)))
    bank.names.flatMap { name =>
      val origin = this.origin.eqTo(name)
      if (origin == FALSE) { Vector.empty } else {
        directions.map { case (direction, guard) =>
          val (state, valid) = bank.neighbors(name)(direction)
          (state, AND(origin, guard), valid)
        }
      }
    }
  }
  def describe(): (Ref, Expr, Expr) = describe(offset)
  def describe(offset: Bits): (Ref, Expr, Expr) = {
    var endpoint = EMPTY
    var live = FALSE
    var valid = FALSE
    for ((state, guard, permitted) <- variants(offset)) {
      endpoint = Ref.select(guard, state.focus, endpoint)
      live = OR(live, AND(guard, state.live))
      valid = OR(valid, AND(guard, permitted))
    }
    (endpoint, live, valid)
  }
  def locationFlags(): (Expr, Expr) = locationFlags(offset)
  /** Availability does not construct discarded endpoint-selection DAGs. */
  def locationFlags(offset: Bits): (Expr, Expr) = {
    var live = FALSE
    var valid = FALSE
    for ((state, guard, permitted) <- variants(offset)) {
      live = OR(live, AND(guard, state.live))
      valid = OR(valid, AND(guard, permitted))
    }
    (live, valid)
  }
  def available(): Expr = {
    val (live, valid) = locationFlags()
    val low = Bits(offset.bits.take(bank.width))
    AND(valid, OR(neg(live), low.unsignedLess(bank.phase)))
  }
  def canMove(amount: BigInt): Expr = {
    val candidate = offset.add(amount)
    val (live, valid) = locationFlags(candidate)
    val low = Bits(candidate.bits.take(bank.width))
    AND(valid, OR(neg(live), low.unsignedLess(bank.phase), low.equal(bank.phase)))
  }
  def move(amount: BigInt, enabled: Expr = TRUE): Unit = {
    if (enabled != FALSE && amount != 0) {
      require(radius + amount.abs < bank.base, "head lineage exceeds the declared finite window")
      bank.circuit.require(canMove(amount), enabled)
      offset = Bits.select(enabled, offset.add(amount), offset)
      radius += amount.abs
    }
  }
  def copyFrom(other: WindowHead, enabled: Expr = TRUE): Unit = {
    require(other.bank eq bank, "head copies require a common stream")
    origin = Value.select(enabled, other.origin, origin)
    offset = Bits.select(enabled, other.offset, offset)
    radius = if (enabled == TRUE) { other.radius } else { radius.max(other.radius) }
  }
  private def integer(value: Any): BigInt = value match {
    case value: Int => BigInt(value)
    case value: Long => BigInt(value)
    case value: BigInt => value
    case _ => throw new IllegalArgumentException("head motion must have a fixed finite distance")
  }
  def moveSelected(amount: Value[?], enabled: Expr = TRUE): Unit = {
    val bound = amount.domain.map(value => integer(value).abs).max
    require(radius + bound < bank.base, "selected head motion exceeds the prepared finite window")
    val before = offset
    var after = offset
    for (delta <- amount.domain if integer(delta) != 0) {
      val guard = AND(enabled, amount.eqTo(delta))
      bank.circuit.require(canMove(integer(delta)), guard)
      after = Bits.select(guard, before.add(integer(delta)), after)
    }
    offset = after
    radius += bound
  }
  def read(enabled: Expr = TRUE): Value[Any] = {
    val c = bank.circuit
    val available = this.available()
    c.require(available, enabled)
    val (end, live, _) = describe()
    val low = Bits(offset.bits.take(bank.width))
    val fullDistance = Bits(low.bits.map(neg))
    val liveDistance = bank.phase.plus(low.negated()).add(-1)
    val distance = Bits.select(live, liveDistance, fullDistance)
    var endpoint = Ref.select(live, NEW, end)
    for ((enabledBit, bit) <- distance.bits.zipWithIndex) {
      endpoint = Ref.select(enabledBit, bank.hop(endpoint, BigInt(1) << bit), endpoint)
    }
    val value = c.get(endpoint, bank.inputKey, c.alphabet, c.alphabet.head)
    Value.select(available, value, Value.constant(None))
  }
  def readRelative(amount: BigInt, enabled: Expr = TRUE): Value[Any] = {
    require(radius + amount.abs < bank.base, "relative read exceeds the prepared head window")
    new WindowHead(bank, name, origin, offset.add(amount), radius).read(enabled)
  }
  def commit(): Unit = {
    val target = bank.states(name)
    for ((source, guard, valid) <- variants()) {
      bank.circuit.require(valid, guard)
      target.copyFrom(source, guard)
    }
    Bits(offset.bits.take(bank.width)).store(bank.circuit, name + ".offset")
    target.commit()
  }
}

/** Bounded motion over blocks of original, unexpanded input.
  * Old block and neighbor zippers are prepared once. Only selected final
  * states are saved; intermediate motion/copies allocate no cells. Every
  * input character belongs to this stream in consecutive transitions.
  */
final class WindowStream(val circuit: Circuit, nameSeq: Seq[String], radius: BigInt, val prefix: String = "stream") {
  require(radius >= 0, "a nonnegative finite movement radius is required")
  val names: Vector[String] = nameSeq.toVector
  require(names.nonEmpty && names.distinct.size == names.size, "distinct finite head names required")
  val width: Int = math.max(1, radius.bitLength)
  val base: BigInt = BigInt(1) << width
  val phaseKey: String = prefix + ".phase"
  val inputKey: String = prefix + ".input"
  private val oldPhase = Bits.load(circuit, phaseKey, width)
  private val completed = AND(oldPhase.bits*)
  val phase: Bits = oldPhase.add(1)
  circuit.put(inputKey, circuit.input(), circuit.alphabet, circuit.alphabet.head)
  val cells: StackPool = new StackPool(circuit, for (name <- names; side <- Vector(".left", ".right")) yield (name + side) -> 1)
  private def pool(layout: (String, Int)*): StackPool = {
    new StackPool(circuit, for (name <- names; (role, count) <- layout) yield s"$name.incoming.$role" -> count)
  }
  private val front = pool("Br" -> 12)
  private val rear = pool("B" -> 1, "B2" -> 1)
  private val reverse = pool("Fr" -> 6)
  val queueCells: Map[String, StackPool] = Map("F" -> front, "WF" -> front, "Br" -> front, "B" -> rear, "WB" -> rear, "B2" -> rear, "Fr" -> reverse)
  val counterCells: Map[String, Map[String, StackPool]] = Map(
    "m" -> Map("pos" -> pool("m.pos" -> 6), "neg" -> pool("m.neg" -> 8)),
    "c" -> Map("pos" -> pool("c.pos" -> 12), "neg" -> pool("c.neg" -> 2)))
  val states: VectorMap[String, BlockState] = VectorMap.from(names.map(name => name -> new BlockState(this, name)))
  private val jumpFields = mutable.LinkedHashMap.empty[BigInt, String]
  def jumps: VectorMap[BigInt, String] = VectorMap.from(jumpFields)
  val neighbors: VectorMap[String, VectorMap[Int, (BlockState, Expr)]] = VectorMap.from(states.map { case (name, state) =>
    state.completeBlock(completed)
    val center = state.clone()
    val before = center.clone()
    val after = center.clone()
    val canLeft = neg(center.left.empty())
    val canRight = neg(center.live)
    before.moveLeft(canLeft)
    after.moveRight(canRight)
    name -> VectorMap(-1 -> (before, canLeft), 0 -> (center, TRUE), 1 -> (after, canRight))
  })
  val heads: VectorMap[String, WindowHead] = VectorMap.from(names.map(name => name -> new WindowHead(this, name)))
  /** An original-input predecessor by a fixed distance, in O(log B) fields. */
  def jump(distance: BigInt): String = {
    require(distance >= 1, "positive original-input hop required")
    if (!jumpFields.contains(distance)) {
      val name = s"$prefix.back$distance"
      jumpFields(distance) = name
      val path = if (distance == 1) { Vector.empty } else if (distance.testBit(0)) {
        Vector.fill(2)(jump(distance / 2))
      } else {
        val half = distance / 2
        Vector(jump(half)) ++ (if (half > 1) { Vector(jump(half - 1)) } else { Vector.empty })
      }
      circuit.pointers(name) = pointer(path)
    }
    jumpFields(distance)
  }
  def hop(target: Ref, distance: BigInt): Ref = {
    val fromNew = if (distance == 1) { pointer(Vector.empty) } else { pointer(Vector(jump(distance - 1))) }
    Ref(FALSE, choosePointer(target.isNew, fromNew, edge(target.prior, jump(distance))))
  }
  def selectHead(which: Value[?]): WindowHead = {
    val cases = heads.toVector.map { case (name, head) => (which.eqTo(name), head) }.filter(_._1 != FALSE)
    val origin = Value.encoded(names, Vector.tabulate(ScaffoldCircuit.bitWidth(names.size))(bit =>
      OR(cases.map { case (guard, head) => AND(guard, head.origin.bits(bit)) }*)), OR(cases.map(_._1)*))
    val offset = Bits(Vector.tabulate(width + 2)(bit => OR(cases.map { case (guard, head) => AND(guard, head.offset.bits(bit)) }*)))
    new WindowHead(this, "", origin, offset, cases.map(_._2.radius).maxOption.getOrElse(BigInt(0)))
  }
  def commit(): Unit = {
    heads.values.foreach(_.commit())
    phase.store(circuit, phaseKey)
  }
}
